module3/lessons/testing_tools_for_api_consumption.md

```diff
@@ -0,0 +1,468 @@
+end
+```
+
+In the first line of the block, we tell VCR where we want to store the the `cassettes`. We are making use of the `spec/fixtures` folder we already created.
+
+The second line tells VCR what library it should use for intercepting these requests, which will be WebMock. So we are still using WebMock, but VCR is adding additional functionality for recording responses.
+
+Let's reset our test in *spec/requests/api/v1/images/image_request_spec.rb* by removing the lines we added to use webmock.
+
+Now, let’s run our test.
+
+```bash
+$ bundle exec rspec spec/requests/api/v1/images/image_request_spec.rb
+```
+
+You’re going to see a big error, but the important part is:
+
+```ruby
+# --- Caused by: ---
+     # VCR::Errors::UnhandledHTTPRequestError:
+     #
+     #
+     #   ================================================================================
+     #   An HTTP request has been made that VCR does not know how to handle:
+     #      GET https://api.pexels.com/v1/search?query=The%20Beatles
+```
+
+This means that it’s working. 
+
+## Stubbing the Request
+
+In order to use VCR, we wrap our test in a `VCR.use_cassette` block:
+
+*spec/requests/api/v1/images/image_request_spec.rb*
+
+```ruby
+require "rails_helper"
+
+RSpec.describe "Images Endpoint" do
+  describe "happy path" do
+    it "can retrieve an image for a specific artist specific artist" do
+      VCR.use_cassette("beatles_artist_query") do
+
+        get "/api/v1/images?artist=The%20Beatles"
+
+        expect(response).to be_successful
+        json = JSON.parse(response.body, symbolize_names: true)
+
+        expect(json[:data][:id]).to be_nil
+        expect(json[:data][:type]).to eq("image")
+        expect(json[:data][:attributes]).to have_key(:image_url)
+        expect(json[:data][:attributes]).to have_key(:photographer)
+        expect(json[:data][:attributes]).to have_key(:photographer_url)
+        expect(json[:data][:attributes]).to have_key(:alt_text)
+      end
+    end
+  end
+end
+```
+
+The string we passed to `use_cassette` is an identifier for the cassette, so it doesn't really matter what you pass it.
+
+Run your tests and they should be passing. If you look under `spec/fixtures/vcr_cassettes` you should see a `.yml` file that contains your recorded response.
+
+## Filtering Sensitive Data
+
+If you look closely in that `.yml` file you can see our API key in there. We will be pushing these cassettes to GitHub, so we don't want the actual API key to be recorded for the same reasons we don't want to hardcode the API key in our code. We will use a VCR option to replace the actual API key with a placeholder. Open up your `rails_helper.rb` and add another line to the VCR configuration:
+
+*spec/rails_helper.rb*
+
+```ruby
+VCR.configure do |config|
+  config.cassette_library_dir = "spec/fixtures/vcr_cassettes"
+  config.hook_into :webmock
+  config.filter_sensitive_data('<CONGRESS_API_KEY>') { Rails.application.credentials.congress[:key] }
```

Comment:
This should be for the PEXELS key! `config.filter_sensitive_data('<PEXELS_API_KEY>') { Rails.application.credentials.pexels[:key] }`

Reply:
darn! Though I got them all! Thanks for the eyeballS!!!!!